queue_job/controllers/main.py

```diff
@@ -0,0 +1,132 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2015-2016 ACSONE SA/NV (<http://acsone.eu>)
+# Copyright 2013-2016 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+import logging
+import traceback
+from cStringIO import StringIO
+
+from psycopg2 import OperationalError
+
+import odoo
+from odoo import _, api, http, tools
+from odoo.service.model import PG_CONCURRENCY_ERRORS_TO_RETRY
+
+from ..job import Job, ENQUEUED
+from ..exception import (NoSuchJobError,
+                         NotReadableJobError,
+                         RetryableJobError,
+                         FailedJobError,
+                         NothingToDoJob)
+
+_logger = logging.getLogger(__name__)
+
+PG_RETRY = 5  # seconds
+
+
+class RunJobController(http.Controller):
+
+    def _load_job(self, env, job_uuid):
+        """ Reload a job from the backend """
+        try:
+            job = Job.load(env, job_uuid)
+        except NoSuchJobError:
+            # just skip it
+            job = None
+        except NotReadableJobError:
+            _logger.exception('Could not read job: %s', job_uuid)
+            raise
+        return job
+
+    def _try_perform_job(self, env, job):
+        """Try to perform the job."""
+
+        # if the job has been manually set to DONE or PENDING,
+        # or if something tries to run a job that is not enqueued
+        # before its execution, stop
+        if job.state != ENQUEUED:
+            _logger.warning('job %s is in state %s '
+                            'instead of enqueued in /runjob',
+                            job.uuid, job.state)
+            return
+
+        # TODO: set_started should be done atomically with
+        #       update queue_job set=state=started
+        #       where state=enqueid and id=
+        job.set_started()
+        job.store()
+        http.request.env.cr.commit()
+
+        _logger.debug('%s started', job)
+        job.perform()
```

Comment:
the sudo is done in `perform()` that executes `self.func()`, the latter being the method call on the sudoed recordset: https://github.com/guewen/queue/blob/43bb875adc421f6773a07ae6ba143014d7a0e600/queue_job/job.py#L419